In [1]:
import numpy as np
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt
import sklearn

from pandas import Series, DataFrame
from pylab import rcParams
from sklearn import preprocessing
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import classification_report
from statsmodels.stats.outliers_influence import variance_inflation_factor  #for VIF
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

import tensorflow
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, BatchNormalization, Input
from tensorflow.keras import optimizers,regularizers

## Task 4
#### Import “Train.csv” into your Jupyter notebook. 

In [3]:
pdata = pd.read_csv('C:\\Users\\stunn\\Notebook\\Capstone\\Train.csv')

In [4]:
pdata.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   2     120000    2          2         2   26     -1      2      0      0   
1   4      50000    2          2         1   37      0      0      0      0   
2   5      50000    1          2         1   57     -1      0     -1      0   
3   6      50000    1          1         2   37      0      0      0      0   
4   7     500000    1          1         2   29      0      0      0      0   

   ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0  ...     3272.0     3455.0     3261.0       0.0    1000.0    1000.0   
1  ...    28314.0    28959.0    29547.0    2000.0    2019.0    1200.0   
2  ...    20940.0    19146.0    19131.0    2000.0   36681.0   10000.0   
3  ...    19394.0    19619.0    20024.0    2500.0    1815.0     657.0   
4  ...   542653.0   483003.0   473944.0   55000.0   40000.0   38000.0   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  Default  
0    1000.0       0.0    2000.0        1  
1    1100.0    1069.0    1000.0        0  
2    9000.0     689.0     679.0        0  
3    1000.0    1000.0     800.0        0  
4   20239.0   13750.0   13770.0        0  

[5 rows x 25 columns]

In [5]:
# Check for missing values if any.
pdata.isnull().values.any()

False

In [6]:
# Convert the relevant variables such as payment variables (Pay0-Pay6 and customer related variables) to categorical variables as appropriate.

payment_columns = ['SEX','EDUCATION','MARRIAGE','PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6']
pdata[payment_columns] = pdata[payment_columns].astype('category')

In [7]:
#pdata.info()

In [8]:
# Dropping irrelevant columns

pdata.drop(['ID', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6','PAY_AMT5'],axis=1,inplace=True)
pdata.head()

LIMIT_BAL SEX EDUCATION MARRIAGE  AGE PAY_0 PAY_2 PAY_3 PAY_4 PAY_5 PAY_6  \
0     120000   2         2        2   26    -1     2     0     0     0     2   
1      50000   2         2        1   37     0     0     0     0     0     0   
2      50000   1         2        1   57    -1     0    -1     0     0     0   
3      50000   1         1        2   37     0     0     0     0     0     0   
4     500000   1         1        2   29     0     0     0     0     0     0   

   PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT6  Default  
0       0.0    1000.0    1000.0    1000.0    2000.0        1  
1    2000.0    2019.0    1200.0    1100.0    1000.0        0  
2    2000.0   36681.0   10000.0    9000.0     679.0        0  
3    2500.0    1815.0     657.0    1000.0     800.0        0  
4   55000.0   40000.0   38000.0   20239.0   13770.0        0

Variable(s) that would cause multicollinearity: 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6'

Additional Variables dropped: ID

In [10]:
X_train = pdata.iloc[:,0:-1].values
y_train = pdata['Default'].values

In [11]:
X_train = preprocessing.normalize(X_train)
y_train = np.array(y_train)

In [12]:
pdataT = pd.read_csv('C:\\Users\\stunn\\Notebook\\Capstone\\Test.csv')
pdataT.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  \
0   5170     140000    2          2         2   26      0      0      2   
1  22727      90000    2          1         2   31      1     -2     -2   
2   4884     150000    2          1         2   32      2      2      2   
3   7142      30000    2          2         2   21     -2     -2     -2   
4  19043      70000    2          2         2   23      2      2      2   

   PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0      0  ...     125046    46364.0      83460    125000       6.0    4578.0   
1     -2  ...          0        0.0          0         0       0.0       0.0   
2      0  ...     123954   131543.0     134319      6000    3000.0    4600.0   
3     -2  ...       4311     1551.0          0       780    3260.0    4311.0   
4      0  ...      18013    19315.0      19859      2850       0.0    1400.0   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  Default  
0    1427.0   87179.0   26291.0        1  
1       0.0       0.0       0.0        0  
2    9800.0    5100.0    5000.0        1  
3    1551.0       0.0       0.0        0  
4    1600.0    1000.0       0.0        1  

[5 rows x 25 columns]

In [13]:
pdataT.drop(['ID', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6','PAY_AMT5'],axis=1,inplace=True)
pdataT.head()

LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0     140000    2          2         2   26      0      0      2      0   
1      90000    2          1         2   31      1     -2     -2     -2   
2     150000    2          1         2   32      2      2      2      0   
3      30000    2          2         2   21     -2     -2     -2     -2   
4      70000    2          2         2   23      2      2      2      0   

   PAY_5  PAY_6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT6  Default  
0      0     -1    125000       6.0    4578.0    1427.0   26291.0        1  
1     -2     -2         0       0.0       0.0       0.0       0.0        0  
2      0      2      6000    3000.0    4600.0    9800.0    5000.0        1  
3     -2     -2       780    3260.0    4311.0    1551.0       0.0        0  
4      0      2      2850       0.0    1400.0    1600.0       0.0        1

In [14]:
X_test = pdataT.iloc[:,:-1].values
y_test = pdataT['Default'].values

In [15]:
X_test = preprocessing.normalize(X_test)
y_test = np.array(y_test)

In [16]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(26000, 16)
(4000, 16)
(26000,)
(4000,)


In [17]:
y_train = tensorflow.keras.utils.to_categorical(y_train, num_classes=2)
y_test = tensorflow.keras.utils.to_categorical(y_test,num_classes= 2)
y_train[0]

array([0., 1.])

In [18]:
model = Sequential()

In [19]:
model.add(Input(shape=(16,)))
model.add(Dense(8, activation='relu', kernel_initializer='normal'))
model.add(Dense(8, activation='relu', kernel_initializer='normal'))
model.add(Dense(2, activation='softmax'))

In [20]:
#model.compile(optimizer = sgd, loss='categorical_crossentropy', metrics=['accuracy'])
model.compile(optimizer = 'adam',loss='categorical_crossentropy')

In [21]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 8)              │           136 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 8)              │            72 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 2)              │            18 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 226 (904.00 B)

 Trainable params: 226 (904.00 B)

 Non-trainable params: 0 (0.00 B)

In [22]:
model.fit(X_train, y_train, epochs = 100, verbose = 1, batch_size=15)

#fit(self, x=None, y=None, batch_size=None, epochs=1, verbose=1, callbacks=None, validation_split=0.0, validation_data=None, shuffle=True, 
#class_weight=None, sample_weight=None, initial_epoch=0, steps_per_epoch=None, validation_steps=None)

Epoch 1/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.5573
Epoch 2/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5242
Epoch 3/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5258
Epoch 4/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5287
Epoch 5/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5214
Epoch 6/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - loss: 0.5288
Epoch 7/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5274
Epoch 8/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5242
Epoch 9/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5280
Epoch 10/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5265
Epoch 11/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5288
Epoch 12/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5304
Epoch 13/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5320
Epoch 14/100
1734/1734 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.5226
Epoch 15/100
17

In [23]:
results = model.evaluate(X_test, y_test)

125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.4878


In [24]:
print(model.metrics_names)
print(results) 

['loss']
0.48560449481010437


In [25]:
Y_pred_cls = model.predict(X_test, verbose=0)
Y_pred_cls

array([[0.85276306, 0.147237  ],
       [0.7628213 , 0.23717868],
       [0.7515917 , 0.24840827],
       ...,
       [0.7519113 , 0.24808872],
       [0.8055446 , 0.19445542],
       [0.7929203 , 0.20707974]], dtype=float32)

In [26]:
YYpred = np.argmax( Y_pred_cls, axis=1)
YYpred

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [27]:
YYtest = np.argmax( y_test, axis=1) 

In [28]:
YYtest

array([1, 0, 1, ..., 0, 1, 0], dtype=int64)

In [29]:
print(metrics.classification_report(YYtest, YYpred))

              precision    recall  f1-score   support

           0       0.79      0.99      0.88      3118
           1       0.59      0.07      0.12       882

    accuracy                           0.78      4000
   macro avg       0.69      0.53      0.50      4000
weighted avg       0.74      0.78      0.71      4000



In [30]:
print(metrics.confusion_matrix(YYtest, YYpred))

[[3076   42]
 [ 822   60]]


Accuracy - 0.78
f1 score - 0.88(excellent) and 0.12(not good)
Recall 1 performed poorly at 0.07.
Recall 0 performed well at 0.99.

Precision 1 performed well at 0.79.
Precision 0 did not perform as well and achieved 0.59